### DISCUSSION TITLE
App hangs when calling API Routes

### DISCUSSION THREAD

**Discussion by HialeahLiam** [vercel/next.js#47892] [Help] [answered]
### Summary

When I run my app in dev mode and navigate to my login page, one of my API Routes gets called and NextJS compiled the route. 

However, it never finishes compiling.

*Page stuck in loading state and api call (loginWithCode) stuck in pending*
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/36059197/229696607-72403316-b6ad-4826-bd2b-be401813958a.png">

*next compiling /api/trpc/[trpc] forever*
<img width="1597" alt="image" src="https://user-images.githubusercontent.com/36059197/229696799-91ae4d31-d3d5-492d-bd41-60621e819ca5.png">

When I try to navigate back to the home page or refresh the page while in this stuck stage, 
my app crashes.

<img width="1440" alt="image" src="https://user-images.githubusercontent.com/36059197/229697018-978bbfc2-3476-44a0-9eed-b3c4ebf45e67.png">

<img width="1597" alt="image" src="https://user-images.githubusercontent.com/36059197/229697045-0a1b1164-27d8-4100-b84f-84f77c1a729e.png">

### Additional information

```js
Error: 'entryOptions.layer' is only allowed when 'experiments.layers' is enabled
```

**Comment by HialeahLiam** [comment, marked as answer]
I found the cause of my issue! It lied in next.config.js.

<img width="636" alt="image" src="https://user-images.githubusercontent.com/36059197/229943779-2274871a-af26-4300-9502-5b9506bb1f92.png">

The issue disappeared when I removed the webpack config option from the `moduleExports` object.
My guess is that the `topLevelAwait` config option is experimental and was messing something up. 

I was even running into the error when running rpm run build!

No errors though as soon as I removed that config option.

An interesting thing is `@next/bundle-analyzer` indirectly fixed the issue as well!

I had added it a couple of weeks ago when I ran into this issue for the first time. It magically disappeared so I didn't pay attention to it again, but it appeared again yesterday when I removed the bundle analyzer plugin.

My guess is that the plugin compiles everything in a way that avoids the error I was getting. Not sure!

**Comment by emassimi** [comment]
Just in case this might help anybody else stumbling on this error, I've simply added `layers: true` to the `config.experiments` object, and that fixed the issue without having to remove `topLevelAwait`.

**Reply by BharathKSunil** [reply]
I was trying to integrate Mongoose to NextJs and the [documentation](https://mongoosejs.com/docs/nextjs.html) told me to add `topLevelAwait: true` and I was getting the same error.
After adding `layers: true` I no longer get this error 🚀

**Reply by jitendra-kr** [reply]
I was facing same problem when using `json_typegen_wasm` in in fireboxtools to [convert JSON to typescript](http://localhost:3000/json-tools/json-to-typescript-interface)
After adding` layers: true` I no longer get this error 🚀
